pkg/image/apis/image/validation/whitelist/whitelister.go

```diff
@@ -0,0 +1,268 @@
+}
+
+func (rw *registryWhitelister) AdmitPullSpec(pullSpec string, transport WhitelistTransport) error {
+	ref, err := imageapi.ParseDockerImageReference(pullSpec)
+	if err != nil {
+		return err
+	}
+	return rw.AdmitDockerImageReference(&ref, transport)
+}
+
+func (rw *registryWhitelister) AdmitDockerImageReference(ref *imageapi.DockerImageReference, transport WhitelistTransport) error {
+	const showMax = 5
+	const fullWhitelistLogLevel = 5
+
+	if rw.pullSpecs.Len() > 0 {
+		if rw.pullSpecs.Has(ref.Exact()) || rw.pullSpecs.Has(ref.DockerClientDefaults().Exact()) || rw.pullSpecs.Has(ref.DaemonMinimal().Exact()) {
+			return nil
+		}
+	}
+
+	if rw.registryHostRetriever != nil {
+		if localRegistry, ok := rw.registryHostRetriever.InternalRegistryHostname(); ok {
+			rw.WhitelistRegistry(localRegistry, WhitelistTransportSecure)
+		}
+	}
+
+	var (
+		host, port string
+		err        error
+	)
+	switch transport {
+	case WhitelistTransportAny:
+		host, port, err = net.SplitHostPort(ref.Registry)
+		if err != nil || len(port) == 0 {
+			host = ref.Registry
+			port = ""
+		}
+		if len(host) == 0 {
+			host, _ = ref.RegistryHostPort(false)
+		}
+	case WhitelistTransportInsecure:
+		host, port = ref.RegistryHostPort(true)
+	default:
+		host, port = ref.RegistryHostPort(false)
+	}
+
+	matchHost := func(h string) bool {
+		for _, hp := range rw.whitelist {
+			if stringsutil.IsWildcardMatch(h, hp.host) {
+				if len(port) == 0 {
+					switch hp.port {
+					case "80", "443", "*":
+						return true
+					default:
+						continue
+					}
+				}
+				if stringsutil.IsWildcardMatch(port, hp.port) {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	switch host {
+	case imageapi.DockerDefaultV1Registry, imageapi.DockerDefaultV2Registry:
+		// try to match plain docker.io first to satisfy `docker.io:*` wildcard
+		if matchHost(imageapi.DockerDefaultRegistry) {
+			return nil
+		}
+		fallthrough
+	default:
+		if matchHost(host) {
+			return nil
+		}
+	}
+
+	hostname := ref.Registry
+	if len(ref.Registry) == 0 {
+		if len(port) > 0 {
+			hostname = net.JoinHostPort(host, port)
+		} else {
+			hostname = host
+		}
+	}
+	whitelist := []string{}
+	full := []string{}
+	for i := 0; i < len(rw.whitelist); i++ {
+		if i < showMax-1 || len(rw.whitelist)-showMax == 0 {
+			whitelist = append(whitelist, fmt.Sprintf("%q", net.JoinHostPort(rw.whitelist[i].host, rw.whitelist[i].port)))
+		} else if i < showMax {
+			whitelist = append(whitelist, fmt.Sprintf("and %d more ...", len(rw.whitelist)-showMax+1))
+		} else if !glog.V(fullWhitelistLogLevel) {
+			break
+		}
+		full = append(full, fmt.Sprintf("%q", net.JoinHostPort(rw.whitelist[i].host, rw.whitelist[i].port)))
+	}
+	glog.V(fullWhitelistLogLevel).Infof("registry %q not allowed by whitelist { %s }", hostname, strings.Join(full, ", "))
+	return fmt.Errorf("registry %q not allowed by whitelist { %s }", hostname, strings.Join(whitelist, ", "))
```

Comment:
Please don't use `{` in error messages, use `message: <comma delimited whitelist>`.  You can do this in a follow up.

I.e. `fmt.Errorf("registry %q is not allowed  by whitelist: %s", hostname, strings.Join(whitelist, ", "))`

Reply:
Reworked as suggested.